### ISSUE TITLE
Interest in adding OCaml code snippets in addition to scala and haskell

### ISSUE THREAD

**Issue by ArulselvanMadhavan** [hmemcpy/milewski-ctfp-pdf#169]
Hello,

Thanks for this wonderful resource. After the support `scala` was added to the book, it just got me thinking whether if it'd be useful to also add OCaml snippets in the book. I just started reading the book and I am also learning OCaml in my free time. So, I am also interested in contributing towards this initiative if it gets enough traction. Anyway, the purpose of this issue is to see if there is any interest.

**Comment by mseri** [comment]
I have been thinking about it as well. I have no time at the moment, but it is on my todo list. I would be happy to review them if you plan to prepare them and need a reviewer

**Comment by Leonidas-from-XIV** [comment]
This sounds very cool. Since I just got the book I might be looking into translating the code as I go along too, but it usually takes me an eternity so I'm happy if someone beats me to it!

**Comment by hmemcpy** [comment]
This is a fantastic idea! There are already 2 efforts I know of translating the snippets to [Kotlin](https://github.com/arrow-kt/Category-Theory-for-Programmers.kt) and [Swift](https://github.com/bow-swift/Category-Theory-for-Programmers). An OCaml edition would be most welcomed, judging by the [tweet](https://twitter.com/hmemcpy/status/1080545340976218113) :)

That said, I don't really know OCaml, so my advice would be setting up a repo and do the translations there. Hopefully there exists a tool for OCaml similar to Scala's `tut` (or Kotlin's `ank`) that can perform compile-time verification of code snippets in markdown files. Then, you could create markdown files with both code snippets (Haskell and OCaml), and make sure they typecheck.

I'd be happy to accept a PR (or, at least, a ping) when such effort is underway. Good luck! :)

**Comment by XVilka** [comment]
@hmemcpy yes, there is - see [mdx](https://github.com/realworldocaml/mdx) which is used in Real World OCaml book.

**Comment by ArulselvanMadhavan** [comment]
Thanks @hmemcpy for the tweet and suggestions on the next step. I will start working on this project soon and update this thread.

**Comment by ArulselvanMadhavan** [comment]
@mseri I have started my work on this. I have the snippets for chapter 1. I'd appreciate if you could review it and let me know if something could be done in a better way so that I can correct them sooner than later.

cc:@hmemcpy

**Comment by mseri** [comment]
@AruselvanMadhavan sure, where should I look?

**Comment by ArulselvanMadhavan** [comment]
@mseri - I added the link to my earlier comment a little late. Here it is - https://github.com/ArulselvanMadhavan/ocaml-ctfp/blob/master/chapter1/chapter1.md

**Comment by ArulselvanMadhavan** [comment]
Hi @mseri, When you have some time, can you review the snippets for chapters 2 - 5?

**Comment by mseri** [comment]
Done :)

**Comment by ArulselvanMadhavan** [comment]
@mseri I have the chapter 6 ready for review, when you have some time.

**Comment by XVilka** [comment]
@ArulselvanMadhavan maybe at this point it worth to send a draft/WIP PR here? Will give your project more visibility.

**Comment by ArulselvanMadhavan** [comment]
@XVilka Sure. I'd like to get more eyes on my work as well. @hmemcpy Is it Ok if I submit a WIP PR to this repo? I have snippets upto chapter 7.1

**Comment by ArulselvanMadhavan** [comment]
@mseri Chapter 7 is also ready for review.

**Comment by mseri** [comment]
I have started the 6, I will go on this weekend and hopefully send you some updated soon.
I think @XVilka suggestion is not bad, it is even easier to give feedback on a pull request

**Comment by hmemcpy** [comment]
Looking good! I admit I haven't touched the book in a while, I need to see if my setup still works 😅

Thank you for the effort! I'll look into it.

**Comment by ArulselvanMadhavan** [comment]
@mseri Chapter 8 is also ready for review, when you have time. 
https://github.com/ArulselvanMadhavan/ocaml-ctfp/blob/master/chapter8/README.md

Since my branch is not fork of this repo, I am not able to submit a PR from my repo. I don't know how to build the book from this repo but I am working on figuring it out.

**Comment by ArulselvanMadhavan** [comment]
@mseri FYI. Chapter 10 is also ready for review.

**Comment by mseri** [comment]
Thanks, I'll try to have a look as soon as possible!

**Comment by XVilka** [comment]
@ArulselvanMadhavan you might also find this information useful http://blog.shaynefletcher.org/2017/05/more-type-classes-in-ocaml.html and Monads, Applicatives, etc from Base library https://ocaml.janestreet.com/ocaml-core/latest/doc/base/Base/

**Comment by XVilka** [comment]
@ArulselvanMadhavan This discussion might also interest you with regard to the type classes https://discuss.ocaml.org/t/modular-implicits/144/61

**Comment by rtviii** [comment]
Thanks for your work @ArulselvanMadhavan
